File: 프로그래머스/unrated/136798. 기사단원의 무기/기사단원의 무기.js

function solution(number, limit, power) {
  const arr = new Array(number).fill(0).map((_, i) => i + 1);
  const attackPowerList = arr.map((a) => getDivisors(a));

  return attackPowerList.reduce((totalWeight, eachPower) => {
    if (eachPower > limit) totalWeight += power;
    else totalWeight += eachPower;
    return totalWeight;
  }, 0);
}

const getDivisors = (num) => {
  const divisors = [];
  for (let i = 1; i <= Math.sqrt(num); i++) {
    if (num % i === 0) {
      divisors.push(i);
      if (num / i != i) divisors.push(num / i);
    }
  }

  return divisors.length;
};

File: 프로그래머스/1/136798. 기사단원의 무기/기사단원의 무기.py
import math

def func(n):
    if n == 1:
        return 1
    if n == 2:
        return 2
        
    else:
        c = 0
        for i in range(1, int(math.sqrt(n)) + 1):
            d = n / i
            f_d = math.floor(d)
            if d == f_d:
                if i < f_d:
                    c += 2
                else:
                    c += 1
        return c
   
def solution(number, limit, power):
    s = 0
    for i in range(1, number + 1):
        d = func(i)
        
        if d > limit:
            s += power
        else:
            s += d
    return s

